torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,344 @@
+            "`overlap_with_ddp=True` to use this hook properly"
+        )
+
+    # NOTE: Gloo may hang with this overlapping approach, so we require
+    # NCCL backend for now
+    if distributed_c10d.get_backend() != distributed_c10d.Backend.NCCL:
+        raise RuntimeError(
+            "Overlapping DDP with ZeRO using this approach currently requires "
+            "NCCL backend to avoid hangs"
+        )
+
+    def hook_with_zero_fn(
+        state: Any,
+        bucket: dist.GradBucket,
+    ) -> torch.futures.Future:
+        r"""
+        Returns a :class:`Future` that gives a gradient bucket tensor and
+        performs the equivalent of a :class:`ZeroRedundancyOptimizer`
+        :meth:`step` if ``bucket`` is the last gradient bucket.
+
+        The function performs additional computation on the iteration that
+        the :class:`DistributedDataParallel` buckets are rebuilt to collect
+        information used to implement the modified hook.
+
+        Arguments:
+            state (Any): any state for the hook.
+            bucket (dist.GradBucket): the :class:`DistributedDataParallel`
+                gradient bucket.
+        """
+        fut = hook(state, bucket)
+        overlap_info = zero._overlap_info
+        bucket_index = bucket.get_index()
+
+        # Proceed as normal until the DDP buckets have been rebuilt
+        if not ddp._has_rebuilt_buckets:
+            assert overlap_info.status == _OverlapStatus.UNINITIALIZED
+            return fut
+
+        if overlap_info.status == _OverlapStatus.UNINITIALIZED:
+            overlap_info.status = _OverlapStatus.DDP_HAS_REBUILT_BUCKETS
+
+        rank = zero.global_rank
+        rank_to_update = zero._ddp_bucket_index_to_rank(bucket_index)
+
+        # Once DDP buckets have been rebuilt but ZeRO has not been
+        # properly initialized yet, collect the information needed
+        if overlap_info.status == _OverlapStatus.DDP_HAS_REBUILT_BUCKETS:
+            bucket_params = bucket.get_model_params_for_bucket()
+            assert len(bucket_params) > 0, "Empty bucket"
+            params_per_rank = overlap_info.params_per_rank
+            params_per_bucket = overlap_info.params_per_bucket
+            if rank_to_update == rank:
+                overlap_info.offsets[bucket_index] = len(params_per_rank[rank_to_update])
+            params_per_rank[rank_to_update].extend(bucket_params)
+            params_per_bucket.append(bucket_params)
+            return fut
+
+        assert overlap_info.status == _OverlapStatus.INITIALIZED
+
+        # Save the bucket reference and all-reduce future for the final bucket
+        if rank_to_update == rank:
+            overlap_info.bucket_index_to_bucket[bucket_index] = bucket
+            overlap_info.bucket_index_to_future[bucket_index] = fut
+
+        num_buckets = len(overlap_info.params_per_bucket)
+        # NOTE: The implementation from this point forward assumes that the
```

Comment:
Shall we add a check here to error out if this assumption was violated? (maybe by remembering the last bucket index)

Reply:
This is a good idea. Currently, I am a bit uncertain on how to do this. I will address it in a follow-up PR.

Remembering the last bucket index might be circular since the way we currently tell if it is the last one is by its index. If the buckets fire in order, then we should be guaranteed that the CPU part of each hook executes in order right? In that case, we can reuse the `bucket_indices_seen` data structures used in the other hook and make sure that the current bucket index is 1 more than the previous index seen (clearing it after each iteration).